homeassistant/components/splunk/__init__.py

```diff
@@ -95,18 +117,62 @@ def splunk_event_listener(event):
         except ValueError:
             _state = state.state
 
-        json_body = [
-            {
+        payload = {
+            "time": event.time_fired.timestamp(),
+            "host": name,
+            "event": {
                 "domain": state.domain,
                 "entity_id": state.object_id,
                 "attributes": dict(state.attributes),
-                "time": str(event.time_fired),
                 "value": _state,
-                "host": name,
-            }
-        ]
-
-        post_request(event_collector, json_body, headers, verify_ssl)
+            },
+        }
+        batch.append(json.dumps(payload, cls=JSONEncoder))
+
+        # Enforce only one loop is running
+        if not post_in_progress:
```

Comment:
We may need to use an `asyncio.Lock` instead and move the state change event callback to async context. Only schedule the payload sending on the executor with `hass.async_add_executor_job`.

Reply:
I have implemented this. `asyncio.Lock` is very nice, and are very well suited to this purpose. Thank you for enlightening me.

Reply:
I took this a step further and reworked it into a custom library that uses the same batch/lock pattern, but is fully async and uses the home assistant aiohttp session